### PR TITLE
[Lock][Semaphore] Add union types

### PR DESCRIPTION
**Pull request by jderusse**
| Q             | A
| ------------- | ---
| Branch?       | 6.0
| Bug fix?      | no
| New feature?  | yes
| Deprecations? | no
| Tickets       | -
| License       | MIT
| Doc PR        | -

Split of #41424 for `Lock` and `Semaphore`

### CODE REVIEW

File: src/Symfony/Component/Lock/Store/PdoStore.php [hunk 1]
```diff
@@ -76,7 +76,7 @@ class PdoStore implements PersistingStoreInterface
      * @throws InvalidArgumentException When PDO error mode is not PDO::ERRMODE_EXCEPTION
      * @throws InvalidArgumentException When the initial ttl is not valid
      */
-    public function __construct($connOrDsn, array $options = [], float $gcProbability = 0.01, int $initialTtl = 300)
+    public function __construct(\PDO|Connection|string $connOrDsn, array $options = [], float $gcProbability = 0.01, int $initialTtl = 300)
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Lock/Store/PdoStore.php:78]: let's remove the `@param` line, it's even wrong :)

File: src/Symfony/Component/Lock/Store/PostgreSqlStore.php [hunk 2]
```diff
@@ -52,7 +52,7 @@ class PostgreSqlStore implements BlockingSharedLockStoreInterface, BlockingStore
      * @throws InvalidArgumentException When PDO error mode is not PDO::ERRMODE_EXCEPTION
      * @throws InvalidArgumentException When namespace contains invalid characters
      */
-    public function __construct($connOrDsn, array $options = [])
+    public function __construct(\PDO|Connection|string $connOrDsn, array $options = [])
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Lock/Store/PostgreSqlStore.php:54]: here also

File: src/Symfony/Component/Lock/Store/RedisStore.php [hunk 3]
```diff
@@ -37,15 +37,10 @@ class RedisStore implements SharedLockStoreInterface
     private $supportTime;
 
     /**
-     * @param \Redis|\RedisArray|\RedisCluster|RedisProxy|RedisClusterProxy|\Predis\ClientInterface $redisClient
-     * @param float                                                                                 $initialTtl  the expiration delay of locks in seconds
+     * @param float $initialTtl  the expiration delay of locks in seconds
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Lock/Store/RedisStore.php]: double space after param name

File: src/Symfony/Component/Lock/Store/StoreFactory.php [hunk 4]
```diff
@@ -25,17 +25,8 @@
  */
 class StoreFactory
 {
-    /**
-     * @param \Redis|\RedisArray|\RedisCluster|\Predis\ClientInterface|RedisProxy|RedisClusterProxy|\Memcached|\MongoDB\Collection|\PDO|Connection|\Zookeeper|string $connection Connection or DSN or Store short name
-     *
-     * @return PersistingStoreInterface
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Lock/Store/StoreFactory.php:31]: not sure we can add the real return type here

**Review by nicolas-grekas — COMMENTED** [review]
tests are failing right now, can you please have a look?

File: src/Symfony/Component/Lock/Store/RedisStore.php [hunk 5]
```diff
@@ -292,8 +287,6 @@ private function evaluate(string $script, string $resource, array $args)
                 throw new LockStorageException($e->getMessage(), $e->getCode(), $e);
             }
         }
-
-        throw new InvalidArgumentException(sprintf('"%s()" expects being initialized with a Redis, RedisArray, RedisCluster or Predis\ClientInterface, "%s" given.', __METHOD__, get_debug_type($this->redis)));
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Lock/Store/RedisStore.php:296]: corresponding `use` can be removed now

File: src/Symfony/Component/Lock/Store/StoreFactory.php (hunk 4 above)
Comment (jderusse) [reply src/Symfony/Component/Lock/Store/StoreFactory.php:31]: Revert. Why can't we do that?

File: src/Symfony/Component/Lock/Store/RedisStore.php (hunk 5 above)
Comment (jderusse) [reply src/Symfony/Component/Lock/Store/RedisStore.php:296]: done

File: src/Symfony/Component/Lock/Store/StoreFactory.php (hunk 4 above)
Comment (nicolas-grekas) [reply src/Symfony/Component/Lock/Store/StoreFactory.php:31]: adding a return type breaks child classes
we first need a deprecation layer in 5.4

**Comment by nicolas-grekas** [comment]
Thank you @jderusse.

**Pull request merged**